[dynamo][stream]support device-agnostic stream in dynamo and capture stream/event method in fx graph

zejun-chen opened the pull request:
This PR implements 2 things:
1. support the device agnostic stream and runtime APIs captured by the dynamo.
2. support the stream methods(include the event) captured by the dynamo.

zejun-chen committed:
[dynamo] make stream capture in dynamo for cuda and xpu

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
remove redundant import in torch/_dynamo/variables/builder.py

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Add test case for stream support multi device in dynamo

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
fix the name error in test case: test_cuda_stream_context_manager2

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Fix the test case issue

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
1. use hasattr stream_id to check if the value is Stream
2. use target value to get device in StreamContextVariable create
3. refine the test case

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
1. create a new native api: torch.cuda._set_stream
2. use python API in torch/_dynamo/variables/ctx_manager.py

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Use register stream method to expand multi backends for
dynamo to capture stream

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Refine the stream method container design

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
change to use torch._dyanmo.register method for multi backends
to register stream method

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
change function name get_one_method -> get_method_by_device
fix the arguments error

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
add comments and refine asseration for dynamo stream

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
refine the stream container class and squeeze the 5 regster APIs
to one API

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
fix caller error in stream container class
refine the get_method API

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

jansel commented on torch/_dynamo/stream.py:
```diff
@@ -0,0 +1,98 @@
+import warnings
+
+
+def __singleton(cls):
+    instance = {}
+
+    def get_instance(*args, **kwargs):
+        if cls not in instance:
+            instance[cls] = cls(*args, **kwargs)
+        return instance[cls]
+    return get_instance
```
Why is instance a dict?   Won't it only have one member?

If args/kwargs change, the second version will be silently ignored.  This seems error prone.

zejun-chen replied:
Thank you for comment. Here `__singleton` is only used for  `StreamMethodContainer`, so it has one member. Maybe this decorator can be used in somewhere developer needs to implement a singleton class.

When args/kwargs changed, only the initial created instance is returned, here needs a warning, otherwise there will be a silent ignore.

zejun-chen replied:
This decorator has been removed.

jansel commented on torch/_dynamo/stream.py:
I am curious why this needs to be handled in dynamo with all this registration machinery.  Perhaps we could add a base class for `torch.cuda.Stream` that other backends could inherit from?

zejun-chen replied:
Agree with you. We can implement a base stream class for multi backends and define some basic members and methods. These methods can be override by other inherit backends. In dynamo, the stream instance can be captured by identifying it as a instance of base stream, instead of a specific backend stream. This could be better than doing registration. We will change the code according to the idea.

zejun-chen commented:
@pytorchbot label "topic: not user facing"

zejun-chen committed:
make lint happy

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
revert wrong update for third party

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
Hi, @jansel 

About implementing a stream base, we have a few points:
1. Implement a stream base could involve lots of changes in native backend. The `THPStream`(bind for frontend `torch.Stream`, https://github.com/pytorch/pytorch/blob/34ddf08f2737025fb1447070fae3087889b2e8bb/torch/csrc/Stream.h#L8) needs to be expanded for the abstraction of all stream methods and attributes, which current defined in C10 stream(https://github.com/pytorch/pytorch/blob/34ddf08f2737025fb1447070fae3087889b2e8bb/c10/core/Stream.h#L63), and make `CUDAStream` or `THCPStream` to inherit from `THPStream`. This could lead to many changes in native and meanwhile making an stream abstraction based on CUDA stream could not achieve good generality.
2. For capture stream method in dynamo, making a stream base may have few help because it still cannot cover the private stream methods defined in the device specific stream class.

Any comments are welcomed.
Thank you.

zejun-chen committed:
remove the singleton decorator

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

jansel commented:
Some design may be needed for this.  I don't know the details of how streams are implemented.

Perhaps we could just do the class hierarchy on the python side and not the C++ side.

zejun-chen committed:
implement a stream base class for backends to herit from

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
1. rename stream runtime API container
2. implement a stream base class

Signed-off-by: zejun-chen <zejun.chen@intel.com>

zejun-chen committed:
1. capture stream method
2. capture event and event method
3. expand the RuntimeInterfaceContainer
4. add the associated test cases

Signed-off-by: zejun-chen <zejun.chen@intel.com>

zejun-chen commented:
> Some design may be needed for this. I don't know the details of how streams are implemented.
> 
> Perhaps we could just do the class hierarchy on the python side and not the C++ side.

Thank you for suggestion. I implement a stream base abstraction on the python side in this PR. Meanwhile i trace the stream methods into the FX graph.

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

jgong5 commented on torch/_dynamo/stream.py:
```diff
@@ -0,0 +1,89 @@
+        "set_stream_by_id": torch.cuda.set_stream_by_id,
+        "create_event": torch.cuda.Event,
+    }
+    register_runtime_interface("cuda", cuda_runtime_interface)
```
Perhaps we can consider to consolidate the registration with the device registration interface `register_interface_for_device` in this PR: https://github.com/pytorch/pytorch/pull/109486/files#diff-06553e25e48e1d60f3030458bc46d52067d3d0c3eef2d5fcea29f7e8126bd7c9R162?

Both of them are about device interface abstraction, no need to register multiple times?

zejun-chen replied:
OK. It is feasible. These containers can be added into the `DeviceInterface`, then one registration is ok.

zejun-chen committed:
rename the runtime interface container to stream interface container

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

zejun-chen committed:
Based on device interface, move stream/event resgiter into the
device interface

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
remove the Stream/Event class in DeviceInterface because
it is an interface container

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
Hi, @jansel @jgong5 @gujinghui @EikanWang 

I made a pure python StreamBase abstraction class for backends to inherit from, for now it is useful for dynamo to capture device-agnostic stream. 
For calling stream create API, i define the interfaces in the `DeviceInterface`(work as an API container) for backends to override.
Thus, StreamBase is used for public abstraction and `DeviceInterface.Stream` is used for calling. They have a clear division.
We can have a review.

Thank you.

jgong5 commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1461,6 +1462,13 @@ def _unimplemented():
+        if (
+            isinstance(left, StreamVariable)
+            and isinstance(right, StreamVariable)
+            or isinstance(left, EventVariable)
+            and isinstance(right, EventVariable)
```
```suggestion
            or (isinstance(left, EventVariable)
            and isinstance(right, EventVariable))
```

zejun-chen replied:
ok. Fixed. Thank you.

jgong5 reviewed (changes requested):
One nit on the conditional check. Others LGTM.

zejun-chen committed:
1. Remove Event and Stream in DeviceInterface
2. Add check for Event and Stream in inherit class to make sure they are
inherit from EventBase and StreamBase

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

jansel commented on torch/streambase.py:
Let's add an `_` at the start of the filename to indicate this isn't a public API.

zejun-chen replied:
OK. Make sense. Thank you.

jansel commented on torch/cuda/__init__.py:
```diff
@@ -534,6 +534,21 @@ def stream(stream: Optional["torch.cuda.Stream"]) -> StreamContext:
     return StreamContext(stream)
 
 
+def set_stream_by_id(stream_id, device_index, device_type):
```
```suggestion
def _set_stream_by_id(stream_id, device_index, device_type):
```

zejun-chen replied:
Fixed.

zejun-chen committed:
Make streambase.py to be _streambase.py
Make function set_stream_by_id to be _set_stream_by_id

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

jansel reviewed (changes requested):
public API lints failing

zejun-chen committed:
Fix flake8 error in torch/_dynamo/device_interface.py: DeviceInterfaceMeta

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
> public API lints failing

OK. Fixed.

zejun-chen commented:
@pytorchbot merge

zejun-chen committed:
1. Make StreamBase and EventBase to be internal with '_'
2. Remove log warnings

Signed-off-by: Chen, Zejun <zejun.chen@intel.com>

zejun-chen commented:
@pytorchbot merge

gujinghui commented:
@jansel All passed. Could you help merge it? Thanks.

jgong5 commented:
@pytorchbot merge

zejun-chen commented:
@pytorchbot merge

zejun-chen commented:
@pytorchbot merge

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

zejun-chen commented:
![image](https://github.com/pytorch/pytorch/assets/74231238/0f87b0bc-0a26-4249-b8b3-7029d527753c)
cuda train _**pit_b_224**_
How can i reproduce the issue? It looks like the dynamo has no graph break with this PR, then here checking failed.

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

zejun-chen committed:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

Commit:
Merge branch 'main' into zejun/support_device_agnostic_stream_in_dynamo

zejun-chen commented:
@pytorchbot merge
